File: 20.valid-parentheses.java
import java.util.Stack;

/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        String map = "(){}[]";
        Stack<Character> stack = new Stack<Character>();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(' || c == '[' || c == '{' ) {
                stack.push(c);
            } else {
                if (!stack.isEmpty()) {
                    if (c == ')' && stack.pop() == '(') {
                        continue;
                    } else if (c == '}' && stack.pop() == '{' ) {
                        continue;
                    } else if (c == ']' && stack.pop() == '[') {
                        continue;
                    }
                } 
                return false;
            }
        }

        return stack.isEmpty();
    }
}
// @lc code=end

File: 20.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
public:
    // Use stack to handle Parentheses problem is good idea by LIFO sequence
    // if we meet the left parenthesis and push it to stack
    // if we meet the right parenthesis and to see 
    // if the stack has the left parenthesis to see what they can match

    bool isValid(string s) {
        stack<char> st;

        for (auto parenthesis : s) {
            if (parenthesis == '(' || parenthesis == '{' || parenthesis == '[') {
                st.push(parenthesis);
            } else {
                if (!st.empty() && (leftOf(st.top()) == parenthesis)) {
                    st.pop();
                } else {
                    return false;
                }
            }
        }

        return st.empty();
    }

    char leftOf(char parenthesis) {
        if (parenthesis == '(') {
            return ')';
        }

        if (parenthesis == '{') {
            return '}';
        }

        return ']';
    }
};
// @lc code=end

